modules/network/src/main/java/org/apache/ignite/network/internal/netty/NettyServer.java

```diff
@@ -0,0 +1,234 @@
+     * @param messageListener Message listener.
+     * @param serializationRegistry Serialization registry.
+     */
+    public NettyServer(
+        ServerBootstrap bootstrap,
+        int port,
+        Consumer<NettySender> newConnectionListener,
+        BiConsumer<SocketAddress, NetworkMessage> messageListener,
+        MessageSerializationRegistry serializationRegistry
+    ) {
+        this.bootstrap = bootstrap;
+        this.port = port;
+        this.newConnectionListener = newConnectionListener;
+        this.messageListener = messageListener;
+        this.serializationRegistry = serializationRegistry;
+    }
+
+    /**
+     * Starts the server.
+     *
+     * @return Future that resolves when the server is successfully started.
+     */
+    public CompletableFuture<Void> start() {
+        if (serverStartFuture != null)
+            throw new IgniteInternalException("Attempted to start an already started server");
+
+        bootstrap.group(bossGroup, workerGroup)
+            .channel(NioServerSocketChannel.class)
+            .childHandler(new ChannelInitializer<SocketChannel>() {
+                /** {@inheritDoc} */
+                @Override public void initChannel(SocketChannel ch) {
+                    ch.pipeline().addLast(
+                        /*
+                         * Decoder that uses org.apache.ignite.network.internal.MessageReader
+                         * to read chunked data.
+                         */
+                        new InboundDecoder(serializationRegistry),
+                        // Handles decoded NetworkMessages.
+                        new MessageHandler(messageListener),
+                        /*
+                         * Encoder that uses org.apache.ignite.network.internal.MessageWriter
+                         * to write chunked data.
+                         */
+                        new ChunkedWriteHandler()
+                    );
+
+                    newConnectionListener.accept(new NettySender(ch, serializationRegistry));
+                }
+            })
+            /*
+             * The maximum queue length for incoming connection indications (a request to connect) is set
+             * to the backlog parameter. If a connection indication arrives when the queue is full,
+             * the connection is refused.
+             */
+            .option(ChannelOption.SO_BACKLOG, 128)
+            /*
+             * When the keepalive option is set for a TCP socket and no data has been exchanged across the socket
+             * in either direction for 2 hours (NOTE: the actual value is implementation dependent),
+             * TCP automatically sends a keepalive probe to the peer.
+             */
+            .childOption(ChannelOption.SO_KEEPALIVE, true);
+
+        serverStartFuture = new CompletableFuture<>();
+
+        NettyUtils.toChannelCompletableFuture(bootstrap.bind(port)).whenComplete((ch, throwable) -> {
```

Comment:
I have can see multiple problems with this code:
1. It's quite complicated. Some complications come from the fact that you suddenly start to care about possible exceptions inside `shutdownGracefully` inside the error path, while you still don't care about them in the normal path (see the `Shutdown event loops on server stop` comment).
2. `serverCloseFuture` is not initialized if `bind` throws an exception.

Here's my attempt on improving this code:

```
serverCloseFuture = CompletableFuture.allOf(
    NettyUtils.toCompletableFuture(bossGroup.terminationFuture()),
    NettyUtils.toCompletableFuture(workerGroup.terminationFuture())
);

serverStartFuture = NettyUtils.toChannelCompletableFuture(bootstrap.bind(port))
    .thenAccept(ch -> {
        CompletableFuture<Void> channelCloseFuture = NettyUtils.toCompletableFuture(ch.closeFuture())
            // Shutdown event loops on server stop.
            .whenComplete((v, err) -> shutdownEventLoopGroups());

        serverCloseFuture = CompletableFuture.allOf(serverCloseFuture, channelCloseFuture);

        channel = (ServerChannel) ch;
    })
    .whenComplete((v, err) -> {
        if (err != null)
            shutdownEventLoopGroups();
    });

return serverStartFuture;
```

What do you think?

Reply:
We have to wait until all of the event loops are terminated though, so I changed your suggestion a little bit